src/coreclr/jit/helperexpansion.cpp

```diff
@@ -0,0 +1,1043 @@
+    //      fastPathValue = t_threadStaticBlocks[typeIndex]
+    //      if (fastPathValue != nullptr)
+    //          goto fastPathBb;
+    //
+    // fallbackBb (BBJ_ALWAYS):                                         [weight: 0]
+    //      threadStaticBlockBase = HelperCall();
+    //      goto block;
+    //
+    // fastPathBb(BBJ_ALWAYS):                                          [weight: 0.97]
+    //      threadStaticBlockBase = fastPathValue;
+    //
+    // block (...):                                                     [weight: 1.0]
+    //      use(threadStaticBlockBase);
+
+    // maxThreadStaticBlocksCondBB
+    BasicBlock* maxThreadStaticBlocksCondBB = fgNewBBFromTreeAfter(BBJ_COND, prevBb, asgTlsValue, debugInfo);
+
+    fgInsertStmtAfter(maxThreadStaticBlocksCondBB, maxThreadStaticBlocksCondBB->firstStmt(),
+                      fgNewStmtFromTree(maxThreadStaticBlocksCond));
+
+    // threadStaticBlockNullCondBB
+    BasicBlock* threadStaticBlockNullCondBB =
+        fgNewBBFromTreeAfter(BBJ_COND, maxThreadStaticBlocksCondBB, asgThreadStaticBlockBase, debugInfo);
+    fgInsertStmtAfter(threadStaticBlockNullCondBB, threadStaticBlockNullCondBB->firstStmt(),
+                      fgNewStmtFromTree(threadStaticBlockNullCond));
+
+    // fallbackBb
+    GenTree*    asgFallbackValue = gtNewAssignNode(gtClone(threadStaticBlockLcl), call);
+    BasicBlock* fallbackBb =
+        fgNewBBFromTreeAfter(BBJ_ALWAYS, threadStaticBlockNullCondBB, asgFallbackValue, debugInfo, true);
+
+    // fastPathBb
+    GenTree* asgFastPathValue =
+        gtNewAssignNode(gtClone(threadStaticBlockLcl), gtCloneExpr(useThreadStaticBlockBaseLclValue));
+    BasicBlock* fastPathBb = fgNewBBFromTreeAfter(BBJ_ALWAYS, fallbackBb, asgFastPathValue, debugInfo, true);
+
+    //
+    // Update preds in all new blocks
+    //
+    fgRemoveRefPred(block, prevBb);
+    fgAddRefPred(maxThreadStaticBlocksCondBB, prevBb);
+
+    fgAddRefPred(threadStaticBlockNullCondBB, maxThreadStaticBlocksCondBB);
+    fgAddRefPred(fallbackBb, maxThreadStaticBlocksCondBB);
+
+    fgAddRefPred(fastPathBb, threadStaticBlockNullCondBB);
+    fgAddRefPred(fallbackBb, threadStaticBlockNullCondBB);
+
+    fgAddRefPred(block, fastPathBb);
+    fgAddRefPred(block, fallbackBb);
+
+    maxThreadStaticBlocksCondBB->bbJumpDest = fallbackBb;
+    threadStaticBlockNullCondBB->bbJumpDest = fastPathBb;
+    fastPathBb->bbJumpDest                  = block;
+    fallbackBb->bbJumpDest                  = block;
+
+    // Inherit the weights
+    block->inheritWeight(prevBb);
+    maxThreadStaticBlocksCondBB->inheritWeight(prevBb);
+    threadStaticBlockNullCondBB->inheritWeight(prevBb);
+    fastPathBb->inheritWeight(prevBb);
+
+    // fallback will just execute first time
+    fallbackBb->bbSetRunRarely();
+
+    //
+    // Update loop info if loop table is known to be valid
+    //
+    if (optLoopTableValid && prevBb->bbNatLoopNum != BasicBlock::NOT_IN_LOOP)
```

Comment:
nit: presumably you don't need to check any coditions and just set `bbNatLoopNum` everywhere even if it's wrong